File: ruby/tictactoe/lib/tictactoe/services/runner.rb
module Tictactoe
  module Services
    class Runner
      attr_accessor :game

      def initialize
        self.game = Tictactoe::Models::Game.new(3)
      end

      def add_players
        game.add_player("Gagan1", "O")
        game.add_player("Gagan2", "X")
      end

      # This method won't be called through test cases. We'll do our own play.
      # This is just to test one case
      # This starts with player1
      def play
        raise InsufficientPlayersError if game.players.count < 2

        game.play(1,1) # Gagan1 places 0 on the center
        game.play(0,2) # Gagan2 places X on the top right
        game.play(2,0) # Gagan1 places 0 on bottom left
        game.play(2,2) # Gagan2 places X on bottom right
        game.play(0,0) # Gagan1 places 0 on top left
        game.play(0,1) # Gagan2 places X on top center
        game.play(1,0) # Gagan1 places 0 on left middle and wins
      end
    end
  end
end

File: java/tictactoe/src/main/java/com/gagan93jtg/lld_learning/services/Runner.java
package com.gagan93jtg.lld_learning.services;

import com.gagan93jtg.lld_learning.errors.*;
import com.gagan93jtg.lld_learning.models.Game;

public class Runner {

  Game game;

  public Runner() {
    this.game = new Game(3);
  }

  public void addPlayers() throws PlayerInformationMissingError, InvalidSymbolError, MoreThanTwoPlayersNotSupportedError, DuplicateSymbolError {
    game.addPlayer("Gagan1", "O");
    game.addPlayer("Gagan2", "X");
  }

  public void play() throws InsufficientPlayersError, InvalidCoordinatesError, BoardCellOccupiedError {
    if (game.getPlayers().size() < 2) {
      throw new InsufficientPlayersError("Atleast two players must be present to play the game");
    }

    game.play(1, 1); // Gagan1 places 0 on the center
    game.play(0, 2); // Gagan2 places X on the top right
    game.play(2, 0); // Gagan1 places 0 on bottom left
    game.play(2, 2); // Gagan2 places X on bottom right
    game.play(0, 0); // Gagan1 places 0 on top left
    game.play(0, 1); // Gagan2 places X on top center
    game.play(1, 0); // Gagan1 places 0 on left middle and wins
  }

  public static boolean printMatrix() {
    return true;
  }

  public static void main (String...args) throws InsufficientPlayersError, InvalidCoordinatesError, BoardCellOccupiedError, PlayerInformationMissingError, InvalidSymbolError, MoreThanTwoPlayersNotSupportedError, DuplicateSymbolError {
    Runner runner = new Runner();
    runner.addPlayers();
    runner.play();
  }
}
